[GFC] Add infrastructure for fractional track sizing

yulunwu2 opened the pull request:
#### e1c57f01dad7afa18a614e2ac9679bffa9cff75c
<pre>
[GFC] Add infrastructure for fractional track sizing
<a href="https://bugs.webkit.org/show_bug.cgi?id=303669">https://bugs.webkit.org/show_bug.cgi?id=303669</a>
&lt;<a href="https://rdar.apple.com/165942167">rdar://165942167</a>&gt;

Reviewed by Sammy Gill and Sam Weinig.

This PR adds support for collecting flex track information needed for sizing.

* Source/WebCore/layout/formattingContexts/grid/GridTypeAliases.h:
* Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
(WebCore::Layout::FlexTrack::FlexTrack):
(WebCore::Layout::TrackSizingAlgorithm::sizeTracks):
(WebCore::Layout::TrackSizingAlgorithm::collectFlexTracks):
(WebCore::Layout::TrackSizingAlgorithm::hasFlexTracks):
(WebCore::Layout::TrackSizingAlgorithm::totalFlexFactor):
* Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.h:

Canonical link: <a href="https://commits.webkit.org/305983@main">https://commits.webkit.org/305983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/828f0cf2f4a903d27f9d327e75f0c582db8d9d66

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/140018 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/12399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/168/builds/1529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/148161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/93091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/141891 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/13109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/12551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/148161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/93091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/142968 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/13109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/168/builds/1529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/148161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/13109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/168/builds/1529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo-libwebrtc~~](https://ews-build.webkit.org/#/builders/166/builds/8450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/13109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/168/builds/1529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/150951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/12084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/168/builds/1529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/150951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/12097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/12551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/150951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/168/builds/1529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/67089 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21610 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/12127 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/168/builds/1529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/11868 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/75813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/12136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/11914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -27,12 +27,25 @@
+    LayoutUnit baseSize;
+    LayoutUnit growthLimit;
+
+    FlexTrack(size_t index, Style::GridTrackBreadth::Flex factor, LayoutUnit base, LayoutUnit growth)
```
This can be marked constexpr.

yulunwu2 replied:
Done, thanks for the review!

weinig commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -27,12 +27,25 @@
+    LayoutUnit growthLimit;
+
+    FlexTrack(size_t index, Style::GridTrackBreadth::Flex factor, LayoutUnit base, LayoutUnit growth)
+        : trackIndex(index), flexFactor(factor), baseSize(base), growthLimit(growth)
```
The usual style is to put each of these on their own line.

```suggestion
        : trackIndex(index)
        , flexFactor(factor)
        , baseSize(base)
        , growthLimit(growth)
```

yulunwu2 replied:
Done, thanks for the review!

weinig commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -70,7 +91,7 @@ TrackSizes TrackSizingAlgorithm::sizeTracks(const PlacedGridItems&, const TrackS
 
     // Each track has a base size, a <length> which grows throughout the algorithm and
-    // which will eventually be the track’s final size...
+    // which will eventually be the track's final size...
```
What's this change about?

yulunwu2 replied:
unnecessary punctuation style change. reverted.

weinig commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -133,5 +154,41 @@ UnsizedTracks TrackSizingAlgorithm::initializeTrackSizes(const TrackSizingFuncti
+    FlexTracks flexTracks;
+
+    for (size_t trackIndex = 0; trackIndex < unsizedTracks.size(); ++trackIndex) {
+        const auto& track = unsizedTracks[trackIndex];
```
If you need both the index and the value, you can use:

```
for (auto [trackIndex, track] = indexedRange(unsizedTracks))
```

yulunwu2 replied:
Thanks for the review! Done.

weinig commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -133,5 +154,41 @@ UnsizedTracks TrackSizingAlgorithm::initializeTrackSizes(const TrackSizingFuncti
+
+bool TrackSizingAlgorithm::hasFlexTracks(const UnsizedTracks& unsizedTracks)
+{
+    for (const auto& track : unsizedTracks) {
```
This can be written a bit more concisely using std::ranges::any_of.

```c++
return std::ranges::any_of(unsizedTracks, [](auto& track) { return track.trackSizingFunction.max.isFlex(); });
```

yulunwu2 replied:
Done, thanks for the review!

weinig commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.h:
```diff
@@ -39,6 +39,12 @@ class TrackSizingAlgorithm {
+    static FlexTracks collectFlexTracks(const UnsizedTracks&);
+    static bool hasFlexTracks(const UnsizedTracks&);
+    static double totalFlexFactor(const FlexTracks&);
+
```
Extra newline.

yulunwu2 replied:
Done, thanks for the review!

weinig commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -45,6 +58,14 @@ TrackSizes TrackSizingAlgorithm::sizeTracks(const PlacedGridItems&, const TrackS
+    double totalFlex = totalFlexFactor(flexTracks);
+
+    UNUSED_VARIABLE(hasFlexTracksToProcess);
+    UNUSED_VARIABLE(totalFlex);
```
Kinda weird to leave this here unused. Assuming this is very temporary?

yulunwu2 replied:
Yes, this will be used in a follow up PR.

sammygill commented on Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
```diff
@@ -170,7 +170,7 @@ auto computeGridItemRects = [](const PlacedGridItems& placedGridItems, const Bor
 
 // https://drafts.csswg.org/css-grid-1/#layout-algorithm
-std::pair<UsedTrackSizes, GridItemRects> GridLayout::layout(GridFormattingContext::GridLayoutConstraints, UnplacedGridItems& unplacedGridItems)
+std::pair<UsedTrackSizes, GridItemRects> GridLayout::layout(GridFormattingContext::GridLayoutConstraints layoutConstraints, UnplacedGridItems& unplacedGridItems)
```
Can `layoutConstraints` be const here? (I know you didn't add this but maybe it's easy enough to do if we can make it const which I thiiiink it should be?)

yulunwu2 replied:
yup!

sammygill commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -58,8 +73,28 @@ TrackSizes TrackSizingAlgorithm::sizeTracks(const PlacedGridItems&, const TrackS
     // 4. Expand Flexible Tracks
-    auto expandFlexibleTracks = [] {
-        notImplemented();
+    auto expandFlexibleTracks = [&] {
```
imo this logic should live in it's own static helper function like `initializeTrackSizes` so that this function can remain as simple as possible when reading it. my idea was just to have the lambdas here to annotate the code and replace them with dedicated functions as we implemented them.

Also maybe it's better to early return if we don't have any flex tracks just to reduce the nested scope

sammygill commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -58,8 +73,28 @@ TrackSizes TrackSizingAlgorithm::sizeTracks(const PlacedGridItems&, const TrackS
+        if (hasFlexTracks(unsizedTracks)) {
+            auto flexTracks = collectFlexTracks(unsizedTracks);
+            double totalFlex = totalFlexFactor(flexTracks);
+            if (totalFlex == 0.0)
```
Thoughts on?
```suggestion
            if (!totalFlex)
```

sammygill commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -133,5 +168,58 @@ UnsizedTracks TrackSizingAlgorithm::initializeTrackSizes(const TrackSizingFuncti
+    FlexTracks flexTracks;
+
+    for (auto [trackIndex, track] : indexedRange(unsizedTracks)) {
+        const auto& maxTrackSizingFunction = track.trackSizingFunction.max;
```
I think you can remove the const since the argument already has it

yulunwu2 replied:
Done, thanks for the review!

sammygill commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -133,5 +168,58 @@ UnsizedTracks TrackSizingAlgorithm::initializeTrackSizes(const TrackSizingFuncti
+double TrackSizingAlgorithm::totalFlexFactor(const FlexTracks& flexTracks)
+{
+    double total = 0.0;
+    for (const auto& track : flexTracks)
```
I think you can remove the const since `flexTracks` already has it

yulunwu2 replied:
Done, thanks for the review

sammygill commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -133,5 +168,58 @@ UnsizedTracks TrackSizingAlgorithm::initializeTrackSizes(const TrackSizingFuncti
+    });
+}
+
+double TrackSizingAlgorithm::totalFlexFactor(const FlexTracks& flexTracks)
```
It seems like this is supposed to be the  `Let flex factor sum be the sum of the [flex factors](https://drafts.csswg.org/css-grid-1/#grid-template-columns-flex-factor) of the [flexible tracks](https://drafts.csswg.org/css-grid-1/#flexible-tracks).`  portion of the spec is that right? If so should we maybe name this `flexFactorSum` just to match the spec terms?

Also if this is for that portion of the spec should we be doing `std::max(1.0f, total)` to make sure that it is at least 1?

yulunwu2 replied:
there is a special behavior for when the flex factor sum is <1. I think your rename makes sense, landing that. Thanks for the review!

sammygill commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -133,5 +168,58 @@ UnsizedTracks TrackSizingAlgorithm::initializeTrackSizes(const TrackSizingFuncti
+
+// https://drafts.csswg.org/css-grid-1/#algo-find-fr-size
+// https://drafts.csswg.org/css-grid-1/#leftover-space
+std::optional<LayoutUnit> TrackSizingAlgorithm::leftoverSpace(std::optional<LayoutUnit> availableSpace, const UnsizedTracks& unsizedTracks)
```
Why is `availableSpace` optional? Presumably this is supposed to be "space to fill,"` from https://drafts.csswg.org/css-grid-1/#algo-find-fr-size and from my reading of https://drafts.csswg.org/css-grid-1/#algo-flex-tracks I think this should always be definite

yulunwu2 replied:
struct GridLayoutConstraints {
        std::optional<LayoutUnit> inlineAxisAvailableSpace;
        std::optional<LayoutUnit> blockAxisAvailableSpace;
    };

I suppose we could add an assert somewhere in a separate PR and remove the optional tags from the broader codebase.

sammygill replied:
Ah I see! I'm not 100% sure if we can get rid of the optional from `GridLayoutConstraints` but regardless it almost seems like it would be a logic error if this function got called with a `std::nullopt` at least according to the spec. 

One thing you could do is add an `ASSERT` to make sure that never happens and make the code a bit more clear with regards to that being an error state. From an API standpoint it probably is a bit cleaner for the function to not be optional and have the caller pass in the right value but I will leave it up to you since you probably have a bit better of an understanding of where this code is going!

sammygill commented on Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
```diff
@@ -133,5 +168,58 @@ UnsizedTracks TrackSizingAlgorithm::initializeTrackSizes(const TrackSizingFuncti
+
+    auto leftoverSpace = availableSpace.value() - usedSpace;
+    if (leftoverSpace <= LayoutUnit { 0 })
+        return LayoutUnit { 0 };
```
```suggestion
    if (leftoverSpace <= 0_lu)
        return { };
```

yulunwu2 replied:
done, thanks for the review

sammygill commented on Source/WebCore/layout/formattingContexts/grid/GridLayout.cpp:
```diff
@@ -170,7 +170,7 @@ auto computeGridItemRects = [](const PlacedGridItems& placedGridItems, const Bor
 
 // https://drafts.csswg.org/css-grid-1/#layout-algorithm
-std::pair<UsedTrackSizes, GridItemRects> GridLayout::layout(GridFormattingContext::GridLayoutConstraints, UnplacedGridItems& unplacedGridItems)
+std::pair<UsedTrackSizes, GridItemRects> GridLayout::layout(const GridFormattingContext::GridLayoutConstraints layoutConstraints, UnplacedGridItems& unplacedGridItems)
```
Probably should be `const GridFormattingContext::GridLayoutConstraints&`

yulunwu2 committed:
[GFC] Add infrastructure for fractional track sizing
https://bugs.webkit.org/show_bug.cgi?id=303669
<rdar://165942167>

Reviewed by Sammy Gill and Sam Weinig.

This PR adds support for collecting flex track information needed for sizing.

* Source/WebCore/layout/formattingContexts/grid/GridTypeAliases.h:
* Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.cpp:
(WebCore::Layout::FlexTrack::FlexTrack):
(WebCore::Layout::TrackSizingAlgorithm::sizeTracks):
(WebCore::Layout::TrackSizingAlgorithm::collectFlexTracks):
(WebCore::Layout::TrackSizingAlgorithm::hasFlexTracks):
(WebCore::Layout::TrackSizingAlgorithm::totalFlexFactor):
* Source/WebCore/layout/formattingContexts/grid/TrackSizingAlgorithm.h:

Canonical link: https://commits.webkit.org/305983@main
